### PR TITLE
[Bridge]: Multiple Timed Migrators

### PR DESCRIPTION
**Pull request by TuDo1403**
### Description
This PR allow trigger upgrade for 3 following contracts: `Staking`, `RoninTrustedOrganization`, `SlashIndicator` when period change is tracked in `RoninTimedMigrator`.
### Contract changes

The table below shows the following info:
- **Logic**: the logic is changed.
- **ABI**: the ABI is changed.
- **Init data**: new storage field is declared and needs initializing.
- **Dependent**: needs to be changed due to changes in other contracts.

| **Contract name** | **Logic** | **ABI** | **Init data** | **Dependent** |
|-------------------|:---------:|:-------:|:-------------:|:-------------:|
| BridgeTracking    |           |         |               |               |
| GovernanceAdmin   |           |         |               |               |
| Maintenance       |           |         |               |               |
| SlashIndicator    |     x      |         |               |               |
| RoninTrustedOrg |     x      |         |               |               |
| Staking           |      x     |         |               |               |
| StakingVesting    |           |         |               |               |
| ValidatorSet      |     x      |         |               |               |

### Checklist
- [x] I have clearly commented on all the main functions following the [NatSpec Format](https://docs.soliditylang.org/en/v0.8.0/natspec-format.html)
- [x] The box that allows repo maintainers to update this PR is checked
- [x] I tested locally to make sure this feature/fix works

### CODE REVIEW

File: test/foundry/forking/REP-002/NewBridgeForkTest.t.sol [hunk 1]
```diff
@@ -30,6 +30,7 @@ contract NewBridgeForkTest is RoninTest, BridgeManagerUtils, SignatureConsumer {
   uint256 internal constant DEFAULT_REWARD_PER_PERIOD = 1 ether;
   uint256 internal constant DEFAULT_EXPIRY_DURATION = 5 minutes;
   uint256 internal constant DEFAULT_GAS = 500_000;
+  uint256 internal constant FORK_HEIGHT = 19231486;
```
Comment (nxqbao) [inline test/foundry/forking/REP-002/NewBridgeForkTest.t.sol:36]: Please comment what is this block height.

File: test/foundry/forking/REP-002/UnattachBridgeForkTest.t.sol [hunk 2]
```diff
@@ -76,9 +139,21 @@ contract UnattachBridgeForkTest is RoninTest {
...
+      _getProxyImplementation(_slashIndicatorProxy),
+      NotifiedMigrator(payable(_slashIndicatorSwitcher)).NEW_IMPL()
+    );
   }
 
   function _upgradeToVersionSwitcher() internal fromWho(_getProxyAdmin(RONIN_VALIDATOR_SET_CONTRACT)) {
     RONIN_VALIDATOR_SET_CONTRACT.upgradeTo(_versionSwitcher);
+    _roninTrustedOrgProxy.upgradeTo(_roninTrustedOrgSwitcher);
+    _stakingProxy.upgradeTo(_stakingSwitcher);
+    _slashIndicatorProxy.upgradeTo(_slashIndicatorSwitcher);
```
Comment (nxqbao) [inline test/foundry/forking/REP-002/UnattachBridgeForkTest.t.sol:227]: Write a few lines that interact with the upgraded contracts to assure it works. Test on a few sample methods, especially calling the methods that are changes in the upgrade.

**Review by nxqbao — COMMENTED** [review]
LGTM. Please write deploy script `/src/deploy` for these migrator contracts.

File: test/foundry/forking/REP-002/UnattachBridgeForkTest.t.sol (hunk 2 above)
Comment (TuDo1403) [reply test/foundry/forking/REP-002/UnattachBridgeForkTest.t.sol:227]: updated.

File: test/foundry/forking/REP-002/NewBridgeForkTest.t.sol (hunk 1 above)
Comment (TuDo1403) [reply test/foundry/forking/REP-002/NewBridgeForkTest.t.sol:36]: // @dev fork height before REP-002 upgrade.

File: test/foundry/forking/REP-002/UnattachBridgeForkTest.t.sol [hunk 3]
```diff
@@ -54,10 +123,38 @@ contract UnattachBridgeForkTest is RoninTest {
...
+    );
+
+    _updateDuplicatedTrustedOrg(seed);
+    _applyValidatorCandidate(seed);
+
+    vm.warp(block.timestamp + 3 seconds);
+    vm.roll(block.number + 1);
+
+    _fastForwardToNextDay();
+    _wrapUpEpoch();
```
Comment (nxqbao) [inline test/foundry/forking/REP-002/UnattachBridgeForkTest.t.sol]: Please assert the method calls emit events as expected.

Comment (TuDo1403) [reply test/foundry/forking/REP-002/UnattachBridgeForkTest.t.sol]: added.

File: src/deploy/ronin-trusted-organization-notified-migrator.ts
Comment (nxqbao) [inline src/deploy/ronin-trusted-organization-notified-migrator.ts:1]: Please bring these migrator contracts into a sub folder for easier management

File: src/deploy/ronin-trusted-organization-notified-migrator.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  }
+
+  await deploy('NotifiedMigrator', {
+    from: deployer,
+    log: true,
+    args: [proxyDepl.address, prevImplAddr, newImplDelp.address, roninValidatorAddr.address],
+  });
+};
+
+deploy.tags = ['RoninTrustedOrganizationNotifiedMigrator'];
```
Comment (nxqbao) [inline src/deploy/ronin-trusted-organization-notified-migrator.ts]:
```suggestion
deploy.tags = ['RoninTrustedOrganizationNotifiedMigrator', 'MigratorBridgeDetachV0_6'];
```

File: src/deploy/slash-indicator-notified-migrator.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  }
+
+  await deploy('NotifiedMigrator', {
+    from: deployer,
+    log: true,
+    args: [proxyDepl.address, prevImplAddr, newImplDelp.address, roninValidatorAddr.address],
+  });
+};
+
+deploy.tags = ['SlashIndicatorNotifiedMigrator'];
```
Comment (nxqbao) [inline src/deploy/slash-indicator-notified-migrator.ts]:
```suggestion
deploy.tags = ['SlashIndicatorNotifiedMigrator', 'MigratorBridgeDetachV0_6'];
```

File: src/deploy/staking-notified-migrator.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  }
+
+  await deploy('NotifiedMigrator', {
+    from: deployer,
+    log: true,
+    args: [proxyDepl.address, prevImplAddr, newImplDelp.address, roninValidatorAddr.address],
+  });
+};
+
+deploy.tags = ['StakingNotifiedMigrator'];
```
Comment (nxqbao) [inline src/deploy/staking-notified-migrator.ts]:
```suggestion
deploy.tags = ['StakingNotifiedMigrator', 'MigratorBridgeDetachV0_6'];
```

File: src/deploy/slash-indicator-notified-migrator.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  const { deploy } = deployments;
+  const { deployer } = await getNamedAccounts();
+
+  const proxyDepl = await deployments.get('SlashIndicatorProxy');
+  const newImplDelp = await deployments.get('SlashIndicatorLogic');
+
+  const IMPLEMENTATION_SLOT = '0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc';
+  const prevImplRawBytes32 = await ethers.provider.getStorageAt(proxyDepl.address, IMPLEMENTATION_SLOT);
+  const prevImplAddr = '0x' + prevImplRawBytes32.slice(26);
+  const roninValidatorAddr = await deployments.get('RoninValidatorSetProxy');
```
Comment (nxqbao) [inline src/deploy/slash-indicator-notified-migrator.ts]: bring it to a helper to reduce duplicated code

File: src/deploy/ronin-trusted-organization-notified-migrator.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
```
Comment (nxqbao) [inline src/deploy/ronin-trusted-organization-notified-migrator.ts]: I think should put at least a dependency, e.g. the proxy `RoninTrustedOrganizationProxy`. Same for the later two deploy scripts.

File: test/foundry/forking/REP-002/NewBridgeForkTest.t.sol [hunk 9]
```diff
@@ -31,6 +31,9 @@ contract NewBridgeForkTest is RoninTest, BridgeManagerUtils, SignatureConsumer {
   uint256 internal constant DEFAULT_EXPIRY_DURATION = 5 minutes;
   uint256 internal constant DEFAULT_GAS = 500_000;
 
+  // @dev fork height before REP-002 upgrade
+  uint256 internal constant FORK_HEIGHT = 19231486;
```
Comment (nxqbao) [inline test/foundry/forking/REP-002/NewBridgeForkTest.t.sol]: Please add a todo to indicate this must be tested again on mainnet

```suggestion
  // @dev fork height before REP-002 upgrade
  // TODO: must be tested on mainnet
  uint256 internal constant FORK_HEIGHT = 19231486;
```

File: src/deploy/ronin-trusted-organization-notified-migrator.ts [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  console.info('  new impl ', newImplDelp.address);
+  console.info('  prev impl', prevImplAddr);
+  console.info('  ronin validator set', roninValidatorAddr.address);
+
+  if (prevImplAddr == DEFAULT_ADDRESS) {
+    console.error('Invalid prev impl');
+    return;
+  }
+
+  await deploy('NotifiedMigrator', {
```
Comment (nxqbao) [inline src/deploy/ronin-trusted-organization-notified-migrator.ts]: I remember it must declare the deployment name? If not, they will be replaced each other. 

```suggestion
  const deployment = await deploy('RoninValidatorSetNotifiedMigrator', {
    contract: 'NotifiedMigrator',
    ...
```

**Review by nxqbao — CHANGES_REQUESTED** [review]

File: src/deploy/ronin-trusted-organization-notified-migrator.ts (hunk 10 above)
Comment (TuDo1403) [reply src/deploy/ronin-trusted-organization-notified-migrator.ts]: resolved.

File: src/deploy/ronin-trusted-organization-notified-migrator.ts (hunk 8 above)
Comment (TuDo1403) [reply src/deploy/ronin-trusted-organization-notified-migrator.ts]: added.

File: src/deploy/slash-indicator-notified-migrator.ts (hunk 7 above)
Comment (TuDo1403) [reply src/deploy/slash-indicator-notified-migrator.ts]: updated.

File: src/deploy/ronin-trusted-organization-notified-migrator.ts
Comment (TuDo1403) [reply src/deploy/ronin-trusted-organization-notified-migrator.ts:1]: updated.

File: src/utils.ts [hunk 11]
```diff
@@ -40,6 +40,12 @@ export const randomAddress = () => {
   return new ethers.Wallet(ethers.utils.randomBytes(32)).address;
 };
 
+export const getImplementOfProxy = async (address: Address): Promise<string> => {
+  const IMPLEMENTATION_SLOT = '0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc';
+  const implRawBytes32 = await ethers.provider.getStorageAt(address, IMPLEMENTATION_SLOT);
+  return '0x' + implRawBytes32.slice(-40);
```
Comment (nxqbao) [inline src/utils.ts:46]: nice trick

**Review by nxqbao — APPROVED** [review]
LTGM

**Pull request merged**